### PR TITLE
[core] Introduce metadata.stats-dense-store to reduce meta size for multiple columns table

### PR DESCRIPTION
**Pull request by JingsongLi**
### Purpose

At present, Paimon uses `BinaryRow` to store statistical information. Generally, it is not a problem, but some businesses have fields with over 3000 columns.

The `BinaryRow` structure has a characteristic that each field occupies a fixed 8 bytes, and for a 3000 column table, one `BinaryRow` has 25kb of storage. `SimpleStats` has 3 `BinaryRow`, then it will be 100 kb storage. 100000 files have GB level storage. This is unacceptable.

This PR：
1. Introduce `metadata.stats-dense-store` to a dense mode to store `SimpleStats` and `valueStatsCols` in `DataFileMeta`.
2. You can set `metadata.stats-mode` = `none`, then `valueStatsCols` will be empty list, and `SimpleStats` is empty.
3. You can also set `fields.b.stats-mode` = `full` to enable stats for specific columns to enable data skipping, the meta storage will only contain `b` column.

### Tests

- `AppendOnlyFileStoreTableTest`
- `PrimaryKeyFileStoreTableTest`

### CODE REVIEW

**Comment by wwj6591812** [comment]
Good，we just encountered a similar problem today!

File: paimon-core/src/main/java/org/apache/paimon/stats/SimpleStatsEvolution.java [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+        }
+    }
+
+    private static class NullCountsEvoArray implements InternalArray {
+
+        private final int[] indexMapping;
+        private final InternalArray array;
+        private final long notFoundValue;
+
+        protected NullCountsEvoArray(int[] indexMapping, InternalArray array, long notFoundValue) {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/stats/SimpleStatsEvolution.java:126]: Why protected?

Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/stats/SimpleStatsEvolution.java:126]: Why protected?

Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/stats/SimpleStatsEvolution.java:126]: Why protected?

File: paimon-core/src/main/java/org/apache/paimon/table/sink/CommitMessageSerializer.java [hunk 2]
```diff
@@ -104,18 +107,20 @@ public List<CommitMessage> deserializeList(int version, DataInputView view) thro
...
-                            dataFileSerializer.deserializeList(view)),
+                            fileDeserializer.get(), fileDeserializer.get(), fileDeserializer.get()),
                     new CompactIncrement(
-                            dataFileSerializer.deserializeList(view),
-                            dataFileSerializer.deserializeList(view),
-                            dataFileSerializer.deserializeList(view)),
+                            fileDeserializer.get(), fileDeserializer.get(), fileDeserializer.get()),
                     new IndexIncrement(
                             indexEntrySerializer.deserializeList(view),
                             indexEntrySerializer.deserializeList(view)));
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/sink/CommitMessageSerializer.java:126]: Condition 'version <= 2' is always 'true'

File: paimon-core/src/main/java/org/apache/paimon/table/source/DataSplit.java [hunk 3]
```diff
@@ -316,6 +317,9 @@ private static FunctionWithIOException<DataInputView, DataFileMeta> getFileMetaS
             DataFileMeta08Serializer serializer = new DataFileMeta08Serializer();
             return serializer::deserialize;
         } else if (version == 2) {
+            DataFileMeta09Serializer serializer = new DataFileMeta09Serializer();
+            return serializer::deserialize;
+        } else if (version == 3) {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/source/DataSplit.java:322]: public static final int DATA_FILE_META_VERSION_1 = 1;
public static final int DATA_FILE_META_VERSION_2= 2;
public static final int DATA_FILE_META_VERSION_3= 3;

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java [hunk 4]
```diff
@@ -1026,7 +1026,7 @@ public class CoreOptions implements Serializable {
     public static final String STATS_MODE_SUFFIX = "stats-mode";
 
     public static final ConfigOption<String> METADATA_STATS_MODE =
-            key("metadata." + STATS_MODE_SUFFIX)
+            key("metadata.stats-mode")
```
Comment (wwj6591812) [inline paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:1029]: Why do this?

File: paimon-core/src/test/java/org/apache/paimon/manifest/ManifestCommittableSerializerCompatibilityTest.java [hunk 5]
```diff
@@ -105,6 +106,76 @@ public void testProduction() throws IOException {
...
+                        3,
+                        Arrays.asList("extra1", "extra2"),
+                        Timestamp.fromLocalDateTime(LocalDateTime.parse("2022-03-02T20:20:12")),
+                        11L,
+                        new byte[] {1, 2, 4},
+                        FileSource.COMPACT,
+                        null);
+        List<DataFileMeta> dataFiles = Collections.singletonList(dataFile);
+
+        LinkedHashMap<String, Pair<Integer, Integer>> dvRanges = new LinkedHashMap<>();
```
Comment (wwj6591812) [inline paimon-core/src/test/java/org/apache/paimon/manifest/ManifestCommittableSerializerCompatibilityTest.java:142]: Map<String, Pair<Integer, Integer>> dvRanges = new LinkedHashMap<>();

**Comment by wwj6591812** [comment]
Please modify this doc : https://paimon.apache.org/docs/master/flink/sql-ddl/#specify-statistics-mode

**Comment by wwj6591812** [comment]
A question : 
If I have a paimon table with 1000 columns, and I just want field B to use dense stat, other columns close stats.
Is my configuration correct？

metadata.stats-dense-store=true
metadata.stats-mode=none
fields.b.stats-mode=truncate(16)

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java [hunk 6]
```diff
@@ -1053,6 +1053,22 @@ public class CoreOptions implements Serializable {
                                                             + STATS_MODE_SUFFIX))
                                     .build());
 
+    public static final ConfigOption<Boolean> METADATA_STATS_DENSE_STORE =
+            key("metadata.stats-dense-store")
+                    .booleanType()
+                    .defaultValue(false)
```
Comment (wwj6591812) [inline paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:1059]: Why the default value is not true?
You are worry about that many users are using old versions of Paimon sdk in their reading engine?

File: paimon-common/src/main/java/org/apache/paimon/table/SystemFields.java [hunk 7]
```diff
@@ -55,4 +55,8 @@ public class SystemFields {
     public static boolean isSystemField(int fieldId) {
         return fieldId >= SYSTEM_FIELD_ID_START;
     }
+
+    public static boolean isSystemField(String field) {
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply paimon-common/src/main/java/org/apache/paimon/table/SystemFields.java:59]: What to solve?

File: paimon-core/src/main/java/org/apache/paimon/table/source/DataSplit.java (hunk 3 above)
Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/table/source/DataSplit.java:322]: No need to do this.

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java (hunk 4 above)
Comment (JingsongLi) [reply paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:1029]: Can become more intuitive in the code.

File: paimon-core/src/test/java/org/apache/paimon/manifest/ManifestCommittableSerializerCompatibilityTest.java (hunk 5 above)
Comment (JingsongLi) [reply paimon-core/src/test/java/org/apache/paimon/manifest/ManifestCommittableSerializerCompatibilityTest.java:142]: IndexFileMeta requires a LinkedHashMap

**Comment by JingsongLi** [comment]
> A question : If I have a paimon table with 1000 columns, and I just want field B to use dense stat, other columns close stats. Is my configuration correct？
> 
> metadata.stats-dense-store=true metadata.stats-mode=none fields.b.stats-mode=truncate(16)

Correct!

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java (hunk 6 above)
Comment (JingsongLi) [reply paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:1059]: Yes, it is true

File: paimon-common/src/main/java/org/apache/paimon/table/SystemFields.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (wwj6591812) [reply paimon-common/src/main/java/org/apache/paimon/table/SystemFields.java:59]: If KEY_FIELD_PREFIX is not in SYSTEM_FIELD_NAMES, then the funtion name "isSystemField" is inappropriate.

File: paimon-core/src/main/java/org/apache/paimon/table/source/DataSplit.java (hunk 3 above)
Comment (wwj6591812) [reply paimon-core/src/main/java/org/apache/paimon/table/source/DataSplit.java:322]: OK

**Comment by wwj6591812** [comment]
+1

**Comment by wwj6591812** [comment]
Looks good to me!
I can't wait to use this feature quickly to solve our online produce problems!

+1

File: paimon-common/src/main/java/org/apache/paimon/table/SystemFields.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (JingsongLi) [reply paimon-common/src/main/java/org/apache/paimon/table/SystemFields.java:59]: So add KEY_FIELD_PREFIX to SYSTEM_FIELD_NAMES, what problem can be solved? Can you write a ut demo?

[the remark this answers is not in the document]
Comment (JingsongLi) [reply paimon-common/src/main/java/org/apache/paimon/table/SystemFields.java:59]: A solution is using a `SYSTEM_FIELD_PREFIXS`, and always using starWith. But it is not good for performance.

Let it go now.

**Comment by LinMingQiang** [comment]
It should be related to this change : https://github.com/apache/paimon/issues/5035

**Pull request merged**